### ISSUE TITLE
Native API: attempting to update metadata of a dataset using JSON is confusing and poorly documented, easily resulting in 500 errors (Internal Server Error) and two cases of a NullPointerException

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#3777]
Originally reported at https://groups.google.com/d/msg/dataverse-community/k6_Pddw6xAU/VxOTOiDhAAAJ and just tested on http://phoenix.dataverse.org running 6eb1584 from the develop branch, attempting to update the metadata of a published dataset using JSON throws a 500 "Internal server error" and a NullPointerException.

As of this writing http://guides.dataverse.org/en/4.6.1/api/native-api.html#datasets reads:

> Updates the current draft version of dataset ``$id``. If the dataset does not have an draft version - e.g. when its most recent version is published, a new draft version is created. The invariant is - after a successful call to this command, the dataset has a DRAFT version with the passed data. The request body is a dataset version, in json format. ::
> 
>     PUT http://$SERVER/api/datasets/$id/versions/:draft?key=$apiKey

It doesn't work, at least with the JSON file below. Hopefully it works with some JSON files but this is unknown if it does. Here are the errors:

`curl -H "Content-type:application/json" -X PUT -d @/tmp/demoHarvard.json http://phoenix.dataverse.org/api/datasets/92/versions/:draft?key=$API_TOKEN`

`{"status":"ERROR","code":500,"message":"Internal server error. More details available at the server logs.","incidentId":"71d195e5-eb28-4e01-9b49-d9d7ad666822"}`

```
[2017-04-19T09:30:13.261-0400] [glassfish 4.1] [SEVERE] [] [edu.harvard.iq.dataverse.api.errorhandlers.ServeletExceptionHandler] [tid: _ThreadID=50 _ThreadName=jk-connector(1)] [timeMillis: 1492608613261] [levelValue: 1000] [[
  API internal error 71d195e5-eb28-4e01-9b49-d9d7ad666822: Null Pointer
java.lang.NullPointerException
	at edu.harvard.iq.dataverse.util.json.JsonParser.parseMetadataBlocks(JsonParser.java:340)
	at edu.harvard.iq.dataverse.util.json.JsonParser.parseDatasetVersion(JsonParser.java:314)
	at edu.harvard.iq.dataverse.util.json.JsonParser.parseDatasetVersion(JsonParser.java:241)
	at edu.harvard.iq.dataverse.api.Datasets.updateDraftVersion(Datasets.java:350)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:151)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:171)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:152)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:104)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:387)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:331)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:103)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:297)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:254)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1028)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:372)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:381)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:344)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:221)
	at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1682)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:344)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(RewriteFilter.java:205)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at edu.harvard.iq.dataverse.api.ApiBlockingFilter.doFilter(ApiBlockingFilter.java:162)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at edu.harvard.iq.dataverse.api.ApiRouter.doFilter(ApiRouter.java:30)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at org.apache.catalina.core.ApplicationDispatcher.doInvoke(ApplicationDispatcher.java:873)
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:739)
	at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:575)
	at org.apache.catalina.core.ApplicationDispatcher.doDispatch(ApplicationDispatcher.java:546)
	at org.apache.catalina.core.ApplicationDispatcher.dispatch(ApplicationDispatcher.java:428)
	at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:378)
	at edu.harvard.iq.dataverse.api.ApiRouter.doFilter(ApiRouter.java:34)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:316)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:160)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:734)
	at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
	at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:174)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:734)
	at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
	at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:412)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:282)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:459)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:167)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:201)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:175)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:561)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:565)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:545)
	at java.lang.Thread.run(Thread.java:745)
]]
```

This is the line that's throwing the `NullPointerException` from `edu.harvard.iq.dataverse.util.json.JsonParser.parseMetadataBlocks(JsonParser.java:340)`:

`Set<String> keys = json.keySet();`

Here is the JSON file being used, which may or may not be expected to work. We need to dig into the details and the API Guide should really include a downloadable JSON file to use as a starting point:

`cat /tmp/demoHarvard.json`

```
{
  "id": 15712,
  "identifier": "FRZAF3",
  "persistentUrl": "http://dx.doi.org/10.5072/FK2/FRZAF3",
  "protocol": "doi",
  "authority": "10.5072/FK2",
  "publisher": "Demo Dataverse",
  "publicationDate": "2017-04-12",
  "datasetVersion": {
    "id": 4204,
    "versionNumber": 1,
    "versionMinorNumber": 0,
    "versionState": "RELEASED",
    "productionDate": "Production Date",
    "lastUpdateTime": "2017-04-12T20:42:20Z",
    "releaseTime": "2017-04-12T20:42:20Z",
    "createTime": "2017-04-12T20:36:12Z",
    "license": "CC0",
    "termsOfUse": "CC0 Waiver",
    "metadataBlocks": {
      "citation": {
        "displayName": "Citation Metadata",
        "fields": [
          {
            "typeName": "title",
            "multiple": false,
            "typeClass": "primitive",
            "value": "test michel"
          },
          {
            "typeName": "subtitle",
            "multiple": false,
            "typeClass": "primitive",
            "value": "Created Using Test Template From Root Dataverse"
          },
          {
            "typeName": "author",
            "multiple": true,
            "typeClass": "compound",
            "value": [
              {
                "authorName": {
                  "typeName": "authorName",
                  "multiple": false,
                  "typeClass": "primitive",
                  "value": "bamouni, bamouni"
                },
                "authorAffiliation": {
                  "typeName": "authorAffiliation",
                  "multiple": false,
                  "typeClass": "primitive",
                  "value": "INRA"
                }
              }
            ]
          },
          {
            "typeName": "datasetContact",
            "multiple": true,
            "typeClass": "compound",
            "value": [
              {
                "datasetContactName": {
                  "typeName": "datasetContactName",
                  "multiple": false,
                  "typeClass": "primitive",
                  "value": "bamouni, bamouni"
                },
                "datasetContactAffiliation": {
                  "typeName": "datasetContactAffiliation",
                  "multiple": false,
                  "typeClass": "primitive",
                  "value": "INRA"
                },
                "datasetContactEmail": {
                  "typeName": "datasetContactEmail",
                  "multiple": false,
                  "typeClass": "primitive",
                  "value": "michel.bamouni@inra.fr"
                }
              }
            ]
          },
          {
            "typeName": "dsDescription",
            "multiple": true,
            "typeClass": "compound",
            "value": [
              {
                "dsDescriptionValue": {
                  "typeName": "dsDescriptionValue",
                  "multiple": false,
                  "typeClass": "primitive",
                  "value": "test"
                }
              }
            ]
          },
          {
            "typeName": "subject",
            "multiple": true,
            "typeClass": "controlledVocabulary",
            "value": [
              "Agricultural Sciences"
            ]
          },
          {
            "typeName": "depositor",
            "multiple": false,
            "typeClass": "primitive",
            "value": "bamouni, bamouni"
          },
          {
            "typeName": "dateOfDeposit",
            "multiple": false,
            "typeClass": "primitive",
            "value": "2017-04-12"
          }
        ]
      }
    },
    "files": [],
    "citation": "bamouni, bamouni, 2017, \"test michel\", doi:10.5072/FK2/FRZAF3, Demo Dataverse, V1"
  }
}
```

Here's how the dataset that I'm attempting to update looks in the GUI. It's published with no files:

![screen shot 2017-04-19 at 9 54 03 am](https://cloud.githubusercontent.com/assets/21006/25183680/47a0358a-24e6-11e7-8b21-2e80699dea43.png)

Tasks for this issue:

- [ ] Come up with a JSON file that works and make it downloadable as an example from the API Guide. In the reporter's words, "It is possible to have an example of working curl command to update my dataset on demo.dataverse.org please?" Fix code on the back end as necessary to make it work.
- [ ] Write a REST Assured test that updates the metadata of a published dataset using JSON to build confidence that this is not a regression.
- [ ] Prevent 500 Internal Server Error from being being displayed to the user. Show a more helpful error message about how they are using the API wrong, if they are using it wrong.
- [ ] Catch NullPointerException
- [ ] Write tests in JsonParserTest.java - Started in 5f0e42e.

**Comment by pdurbin** [comment]
Good news, bad news.

The good news is that there's a workaround but if there are files involved, there's an extra complication and #3413 is related.

The workaround below should be documented in the API Guide. Let's first talk about a dataset without any files to simplify things. It is possible to update a dataset via JSON but you have to know the secret of constructing the JSON file, which I myself had forgotten.

Currently, the best example of how to construct the JSON file is hidden away in my old `edit-dataset-finch1` script that changes the title of a dataset using JSON and the Native API: https://github.com/IQSS/dataverse/blob/v4.6.1/scripts/search/tests/edit-dataset-finch1

I just used it again on my laptop and it works. Here's what it does:

Step 1. Serialize the dataset metadata as JSON using `versions/:latest` and only save the output under `data`:

`curl http://localhost:8080/api/datasets/2837/versions/:latest?key=$API_TOKEN | jq .data > /tmp/issue3777latest.json`

Step 2. Make your edits to the JSON file, changing the title or whatever:

`vim /tmp/issue3777latest.json`

Step 3: Upload the JSON file containing your edits:

`curl -H "Content-type:application/json" -X PUT -d @/tmp/issue3777latest.json http://localhost:8080/api/datasets/2837/versions/:draft?key=$API_TOKEN`

Great! Worth documenting in the API Guide. But what if there are files involved? When you use the `versions/:latest` API to serialize the dataset to JSON, files are included as well. And when you try to upload the JSON file containing your edits... BOOM! NullPointerException:

```
[2017-04-19T13:15:01.507-0400] [glassfish 4.1] [SEVERE] [] [edu.harvard.iq.dataverse.api.errorhandlers.ServeletExceptionHandler] [tid: _ThreadID=104 _ThreadName=http-listener-1(2)] [timeMillis: 1492622101507] [levelValue: 1000] [[
  API internal error 482beccf-73e7-45ee-9118-432b7ae99237: Null Pointer
java.lang.NullPointerException
	at edu.harvard.iq.dataverse.util.json.JsonParser.parseFiles(JsonParser.java:386)
	at edu.harvard.iq.dataverse.util.json.JsonParser.parseDatasetVersion(JsonParser.java:321)
	at edu.harvard.iq.dataverse.api.Datasets.updateDraftVersion(Datasets.java:350)
```

The workaround for files is to delete them from the JSON. Again, this is highly related to #3413. To fix #3413 we should probably drop the `files` section of the JSON on the floor anyway because we really don't want users changing checksum values of their files, which the API currently allows. After #3413 was opened we shipped File Replace, which allows users to change file metadata that we allow such as descriptions and categories. 

I did push a little code while investigating this issue. I added some tests in 5f0e42e which allowed me to better understand the JSON parsing.

I'll also upload the files I'm talking about as well as a tweaked version of the reporter's file, which works for me.

- [demoHarvardTweaked.json.txt](https://github.com/IQSS/dataverse/files/936694/demoHarvardTweaked.json.txt)
- [issue3777latest.json.txt](https://github.com/IQSS/dataverse/files/936697/issue3777latest.json.txt)
- [new.json.txt](https://github.com/IQSS/dataverse/files/936696/new.json.txt) (NullPointerException expected)
- [new-nofiles.json.txt](https://github.com/IQSS/dataverse/files/936695/new-nofiles.json.txt)

**Comment by pdurbin** [comment]
If anyone wants to pick up this issue, here's a branch I created as a starting point: https://github.com/IQSS/dataverse/tree/3777-native-api-update-metadata

**Comment by asconrad** [comment]
The workaround also works for my case, reported in https://groups.google.com/forum/#!msg/dataverse-community/-y4G0s-81EA/bQsWAbmxAwAJ;context-place=forum/dataverse-community.

However with a caveat that may or may not be intended: The new DRAFT version of the updated dataset can only be discovered and retrieved through the graphical interface by clicking the Versions tab on the dataset landing page. It does not seem to be indexed and does not show in the Publication Status: Draft facet or through search. This may be intended behaviour - or a different issue?

**Comment by pdurbin** [comment]
@asconrad that sounds like a separate bug having to do with indexing data into Solr. Can you please create a new issue?

**Comment by asconrad** [comment]
Yes, will do as soon as I am back from a trip/project meeting.
/Anders

Fra: Philip Durbin <notifications@github.com<mailto:notifications@github.com>>
Svar til: IQSS/dataverse <reply@reply.github.com<mailto:reply@reply.github.com>>
Dato: mandag den 15. maj 2017 kl. 20.22
Til: IQSS/dataverse <dataverse@noreply.github.com<mailto:dataverse@noreply.github.com>>
Cc: Anders Conrad <asc@kb.dk<mailto:asc@kb.dk>>, Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Emne: Re: [IQSS/dataverse] Native API: attempting to update metadata of a dataset using JSON is confusing and poorly documented, easily resulting in 500 errors (Internal Server Error) and two cases of a NullPointerException (#3777)


@asconrad<https://github.com/asconrad> that sounds like a separate bug having to do with indexing data into Solr. Can you please create a new issue?

-

**Comment by pdurbin** [comment]
Thanks, @asconrad !

**Comment by pdurbin** [comment]
The same issue was just reported at #3899.

**Comment by pdurbin** [comment]
I mentioned this issue to @rebecabarros today at http://irclog.iq.harvard.edu/dataverse/2017-09-26#i_57794 because she asked how to edit the metadata of a dataset beyond the few "dcterms" fields that are offered by the SWORD API. This issue at least explains the way to do it, even thought it's a little weird and confusing. We should update the documentation.

**Comment by pdurbin** [comment]
While working on #4139 I stubbed out a sample JSON file to update metadata in 445e1ed and added a method to update metadata via our REST Assured API test suite. The API Guide should be updated to make the JSON file a downloadable file.
